### PR TITLE
[RLlib; Pre-checks/better failure behavior] Env Checker for Gym Environments

### LINKED ISSUE #20578: RLlib; Pre-checks/better failure behavior Failure modes environment checking gym environments
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

add checks for gym environments

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by avnishn**
## Why are these changes needed?

Evaluating and checking envs is buried in the codebase, making error messages hard to parse.

I'm going to keep updating this PR, and asking for reviews @sven1977, @richard4912, @gjoliver 
It might be best to filter by most recent commits when looking at this pr.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/pre_checks.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+              VectorEnv]) -> None:
+    """Run pre-checks on env that uncover common errors in environments.
+
+    Args:
+        env: Environment to be checked.
+
+    Raises:
+        ValueError: If env is not one of the supported types.
+    """
+    supported_types = (BaseEnv, gym.Env, MultiAgentEnv, RemoteVectorEnv,
```
Comment (gjoliver) [inline rllib/utils/pre_checks.py:21]: define this as a constant at the top of the file, and use it also in the type annotation?
can we also start a simple unit test file with this PR? basically try to fail whatever tests we are checking here.

**Review by gjoliver — COMMENTED** [review]
why don't you merge the PR while you are moving things here?
I mean, making separate smaller PRs is way better than keep adding to an open PR. Waleed said the optimal PR size is 80 lines :)

File: rllib/utils/pre_checks.py (hunk 1 above)
Comment (avnishn) [reply rllib/utils/pre_checks.py:21]: of course! I will add a constant and add tests here for every one of the checks

File: rllib/utils/pre_checks.py [hunk 2]
```diff
@@ -3,18 +3,17 @@
...
 logger = logging.getLogger(__name__)
 
-SUPPORTED_ENVIRONMENT_TYPES = (BaseEnv, gym.Env, MultiAgentEnv,
-                               RemoteVectorEnv, VectorEnv)
 
-
-def check_env(
-        env: [BaseEnv, gym.Env, MultiAgentEnv, RemoteVectorEnv,
-              VectorEnv]) -> None:
+def check_env(env: ["BaseEnv", "gym.Env", "MultiAgentEnv", "RemoteVectorEnv",
```
Comment (avnishn) [inline rllib/utils/pre_checks.py]: Does anyone know if this will fail sphinx build? 

@sven1977.

I need to do this because I need to local import these packages otherwise I have a circular import issue.

File: rllib/utils/pre_checks.py (hunk 1 above)
Comment (avnishn) [reply rllib/utils/pre_checks.py]: I ended not being able to add this as a constant in this file because of circular imports. Not sure where else I would put it -- maybe in rllib/env/__init__.py but im not sure if thats necessary. Lmk what you think @gjoliver

File: rllib/utils/pre_checks.py (hunk 2 above)
Comment (sven1977) [reply rllib/utils/pre_checks.py]: Can you use `Union["BaseEnv", ...]` instead?
Quotes are fine if you want to avoid circular imports.
What also may help is to use `from typing import TYPE_CHECKING` (search for this in rllib/ to see examples of how imports can be conditioned on whether we are currently type checking or not). But this may not always help.

File: rllib/utils/tests/test_check_env.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class TestGymCheckEnv():
+    def test_has_observation_space(self):
+        env = gym.make("CartPole-v1")
+        del env.observation_space
+        with pytest.raises(AttributeError):
+            check_gym_environments(env)
+
+    def test_has_action_space(self):
+        env = gym.make("CartPole-v1")
+        del env.observation_space
```
Comment (sven1977) [inline rllib/utils/tests/test_check_env.py:18]:
```suggestion
        del env.action_space
```

**Review by sven1977 — APPROVED** [review]
Looks great!
Just a few nits (please keep these in mind for future PRs):
- Use unittest.TestCase for your tests (see other test cases on how we do this).
- sort imports in two blocks a) external and b) ray imports.
- within each import block, sort the individual imports alphabetically.
- Try to avoid yellow highlighted code in PyCharm, e.g.:
```
better:
    assert isinstance(reward, (float, int)), ...
instead of:
    assert isinstance(reward, [float, int]), ...
```

**Comment by sven1977** [comment]
Let's wait for tests to pass, then merge!

**Comment by sven1977** [comment]
Nice!
